pkg/didcomm/messaging/service/http/httpmsgsvc.go

```diff
@@ -19,35 +22,117 @@ const (
 )
 
-// NewHTTPOverDIDComm creates new HTTP over DIDComm message service which serves
-// incoming DIDComm message over HTTP.
-func NewHTTPOverDIDComm() *OverDIDComm {
-	return &OverDIDComm{httpOverDIDComm}
+// RequestHandle handle function for http over did comm message service which gets called by
+// `OverDIDComm` message service to handle incoming request.
+// Args
+// msgID : message ID of incoming message.
+// request: http request derived from incoming DID comm message.
+//
+// Returns
+// error : handle can return error back to service to notify message dispatcher about failures.
+type RequestHandle func(msgID string, request *http.Request) error
+
+// NewDIDCommToHTTPMessageService creates new HTTP over DIDComm message service which serves
+// incoming DIDComm message over HTTP. DIDComm message receiver of [RFC-0351]
+//
+// Reference:
+// https://github.com/hyperledger/aries-rfcs/blob/master/features/0335-http-over-didcomm/README.md
+// https://github.com/hyperledger/aries-rfcs/blob/master/features/0351-purpose-decorator/README.md
+//
+// Args:
+// name: service name, this is mandatory argument.
+//
+// purpose: optional list of purposes to be handled by this http-over-didcomm service [RFC-0351].
+// If not provided only message type will be taken into consideration in acceptance criteria of this message service.
+//
+// httpHandle: handle function to which incoming DIDComm message will be sent after being converted to
+// http request. this is mandatory argument.
+//
+// Returns:
+// OverDIDComm: http over didcomm message service
+// error: arg validation errors
+func NewDIDCommToHTTPMessageService(name string, purpose []string, httpHandle RequestHandle) (*OverDIDComm, error) {
+	if name == "" || httpHandle == nil {
+		return nil, fmt.Errorf(errNameAndHandleMandatory)
+	}
+
+	return &OverDIDComm{
+		name:       name,
+		purpose:    purpose,
+		httpHandle: httpHandle,
+	}, nil
 }
 
 // OverDIDComm is message service which transports incoming DIDComm message over to
 // intended http resource providers
 type OverDIDComm struct {
-	name string
+	name       string
+	purpose    []string
+	httpHandle RequestHandle
 }
 
 // Name of HTTP over DIDComm message service
 func (m *OverDIDComm) Name() string {
 	return m.name
 }
 
-// Accept criteria for HTTP over DIDComm message service
-// TODO purpose based acceptance criteria [Issue #1106]
+// Accept criteria for HTTP over DIDComm message service, it accepts http-didcomm-over message type [RFC-0335]
+// and follows `A tagging system` purpose field validation from RFC-0351
 func (m *OverDIDComm) Accept(msgType string, purpose []string) bool {
-	return msgType == OverDIDCommSpec
+	if msgType != OverDIDCommSpec {
+		return false
+	}
+
+	// if purpose not set, then match only message type.
+	if len(m.purpose) == 0 {
+		return true
+	}
+
+	// match purpose if provided
+	for _, msgPurpose := range purpose {
+		for _, svcPurpose := range m.purpose {
```

Comment:
consider using `sort.Search()` for efficiency

Reply:
`sort.Search()` doesn't directly tell if an element is missing from given array. It just tells about where the item can be inserted in given sorted array. It is not much of use here in this code. And also, purpose array is typically very small so this nested loop doesn't hurt much.

Reply:
This is how you can use `sort.SearchStrings()`:

```go
// prints:
//   contains 'a'? true
//   contains 'b'? false
func main() {
	strings := []string{"d", "e", "c", "a", "g", "f"}	// missing letter 'b'
	sort.Strings(strings)

	fmt.Printf("contains 'a'? %v\n", contains(strings, "a"))
	fmt.Printf("contains 'b'? %v", contains(strings, "b"))
}

func contains(s []string, elm string) bool {
	if idx := sort.SearchStrings(s, elm); idx < len(s) && elm == s[idx] {
		return true
	} else {
		return false
	}
}
```

Reply:
The size of the array is variable and depends on the application.

One of the possible use cases for this framework is to connect *things* (IoT), where you can expect a network of many devices, each tailored to listen for specific events.